Compile/interface/ZJetProduct.h

```diff
@@ -308,17 +308,99 @@ class ZJetProduct : public KappaProduct
+
+    // MPF_jets
+    double GetMPFjets(ZJetSettings const& settings, ZJetEvent const& event, std::string corrLevel) const
+    {
+        if(GetValidJetCount(settings, event, corrLevel) < 2) {
+            return 0;
+        }
+        else
+        {
+            RMFLV jetcontrib;
+            for(uint index = 1; index < GetValidJetCount(settings, event, corrLevel); index++)
+            {
+                KLV* jet = GetValidJet(settings, event, index, corrLevel);
+                if(jet->p4.Pt() < settings.GetMPFSplittingJetPtMin()) break;
+
+                jetcontrib += jet->p4;
+            }
+            return (-1.) * GetTransverseProjectionFraction(&jetcontrib, &m_z.p4);
+        }
+    }
+
+    double GetMPFjets(ZJetSettings const& settings, ZJetEvent const& event) const
+    {
+        return GetMPFjets(settings, event, settings.GetCorrectionLevel());
+    }
+
+    // MPF_unclustered
+    double GetMPFunclustered(ZJetSettings const& settings, ZJetEvent const& event, std::string corrLevel) const
+    {
+        //TODO: #57 implement independent calculation of the unclustered energy contribution
+        return GetMPF(GetMet(settings, event, corrLevel)) - GetMPFlead(settings, event, corrLevel) - GetMPFjets(settings, event, corrLevel);
+    }
+
+    double GetMPFunclustered(ZJetSettings const& settings, ZJetEvent const& event) const
+    {
+        return GetMPFunclustered(settings, event, settings.GetCorrectionLevel());
+    }
+
+    // Helper method for MPF
+    static double GetTransverseProjectionFraction(const RMFLV* lvProj, const RMFLV* lvRef) {
+        double scalPtEt = lvRef->Px() * lvProj->Px() + lvRef->Py() * lvProj->Py();
+        double scalPtSq = lvRef->Px() * lvRef->Px()  + lvRef->Py() * lvRef->Py();
+        return scalPtEt / scalPtSq;
+    }
+
+    // Calculate JNPF
+    double GetJNPF(ZJetSettings const& settings, ZJetEvent const& event, std::string corrLevel) const
+    {
+        if(GetValidJetCount(settings, event, corrLevel) < 1) return DefaultValues::UndefinedDouble;
+        else 
+        {
+            double scalPtEt = 0;
+            double scalPtSqjet = 0;
+            KLV* met = GetMet(settings, event, corrLevel);
+            for(uint index = 0; index < GetValidJetCount(settings, event, corrLevel); index++)
+            {    
+                KLV* jet = GetValidJet(settings, event, index, corrLevel);
+                if(jet->p4.Pt() < settings.GetJNPFJetPtMin()) break;
+                scalPtEt += jet->p4.Px() * met->p4.Px() + jet->p4.Py() * met->p4.Py() + jet->p4.Pz() * met->p4.Pz();
+                scalPtSqjet += jet->p4.Px() * jet->p4.Px()  + jet->p4.Py() * jet->p4.Py() + jet->p4.Pz() * jet->p4.Pz();
+            }
+            double scalPtSqmet = met->p4.Px() * met->p4.Px()  + met->p4.Py() * met->p4.Py() + met->p4.Pz() * met->p4.Pz();
+
+            if(scalPtSqjet == 0 || scalPtSqmet == 0) return DefaultValues::UndefinedDouble;
+            return scalPtEt / (sqrt(scalPtSqjet) * sqrt(scalPtSqmet));
```

Comment:
Equation in 1 is equivalent to 3.:
![image](https://user-images.githubusercontent.com/33454678/86146236-41fd5a80-baf8-11ea-8525-333f1da4c002.png)

Reply:
Mikko answered to my questions: It seems he wants indeed the projection of the pT_jet to pT_Z. Which means, I change remove all Pz components and change MET to pT_Z.

Reply:
What you wrote is indeed equivalent to [2], but not to [3]. The crucial difference is between the magnitude of the vectorial sum and the sum of the magnitudes in the denominator. Consider two back-to-back jets with p1 = -p2 (vectorially). Equation [2] then has (2 |p1|) in the denominator while equation [3] has 0 and is thus undefined.

Reply:
> What you wrote is indeed equivalent to [2], but not to [3]. The crucial difference is between the magnitude of the vectorial sum and the sum of the magnitudes in the denominator. Consider two back-to-back jets with p1 = -p2 (vectorially). Equation [2] then has (2 |p1|) in the denominator while equation [3] has 0 and is thus undefined.

Yes you are right, I thought it is a typo in equation 3. But my implementation in the code really follows equation 2 not 3. So it should be fine.

Reply:
> > What you wrote is indeed equivalent to [2], but not to [3]. The crucial difference is between the magnitude of the vectorial sum and the sum of the magnitudes in the denominator. Consider two back-to-back jets with p1 = -p2 (vectorially). Equation [2] then has (2 |p1|) in the denominator while equation [3] has 0 and is thus undefined.
> 
> Yes you are right, I thought it is a typo in equation 3. But my implementation in the code really follows equation 2 not 3. So it should be fine.

OK, it seems I indeed summarized your implementation wrongly in equation [3]. In reality, the current implementation does not add the jet momenta vectorially as I first thought (sorry about that). But it is still not equivalent to equation [2], since the result is normalized after performing the summation separately in the numerator and the denominator.
However, equation [2] would require normalizing the dot product inside the loop.

Here's some Python code to test this with a few dummy jets and met:

```
import numpy as np

JETS = np.array([[1,2,3],[4,5,3],[9,1,3],[1,3,3]])
MET = np.array([5,2,0])

def jnpf_current_implementation(met, jets):
    scalPtEt = sum(j.dot(met) for j in jets)
    scalPtSqjet = sum(j.dot(j) for j in jets)
    scalPtSqmet = met.dot(met)
    return scalPtEt / np.sqrt(scalPtSqjet) / np.sqrt(scalPtSqmet)

def jnpf_equation_2(met, jets):
    return sum(
        j.dot(met) / np.sqrt(j.dot(j) * met.dot(met))
        for j in jets
    )

def jnpf_equation_3(met, jets):
    sumj = sum(j for j in jets)
    return sumj.dot(met) / np.sqrt(sumj.dot(sumj) * met.dot(met))

print(jnpf_current_implementation(MET, JETS))  # gives 1.36552
print(jnpf_equation_2(MET, JETS))  # gives 2.618
print(jnpf_equation_3(MET, JETS))  # gives 0.813719
```

Reply:
You are right it should sum the `scalPtEt/scalPtSqjet`. I fix this.